### PR TITLE
Design proposal for training

### PR DESCRIPTION
**Pull request by mvsusp**
# Amazon SageMaker Containers

**SageMaker Containers** is the open source library that the SageMaker team uses to create the open source containers
that run on [Amazon SageMaker](https://aws.amazon.com/documentation/sagemaker/):

- https://github.com/aws/sagemaker-tensorflow-containers
- https://github.com/aws/sagemaker-mxnet-containers
    
 You can use SageMaker Containers as a **library** with helper functions for your training script or to 
 **create** your own ML framework container compatible with SageMaker. 
 
### Installing from source
 ```bash
pip install -U .
```

### Using it as a **library**
```sagemaker_containers.training_environment``` contains information about the training container including: 
hyperparameters, instance information (number of CPUs, GPUs, host name), input data information, environment
variables, etc.

Let's suppose that you want to train the Keras script below in SageMaker:
```my_training_script.py```
```python
import argparse
import os

import numpy as np
import tensorflow as tf

parser = argparse.ArgumentParser()
parser.add_argument('--training-data-dir', type=str)
parser.add_argument('--batch-size', type=int)
parser.add_argument('--model-dir', type=str)

args = parser.parse_args()

data = np.load(os.path.join(args.training_data_dir, 'training_data.npz'))
x_train, y_train = data['features'], keras.utils.to_categorical(data['labels'])

model = tf.keras.models.Sequential()
model.add(tf.keras.layers.Dense(64, activation='relu', input_dim=20))
model.add(tf.keras.layers.Dropout(0.5))

sgd = tf.keras.optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])

model.fit(x_train, y_train, epochs=20, batch_size=args.batch_size)

# saves the model in the end of training
model.save(os.path.join(args.model_dir, 'saved_model.h5'))
```

To be able to train the script above in SageMaker, you need to provide the parameters ```--training-data-dir```,
```--batch-size```, ```--model-dir```. ```training_environment``` makes this process very easy:

```python
import sagemaker_containers.training_environment as env
...

# TODO (mvsusp) - create env.training_data_dir and env.evaluation_data_dir helpers
parser.add_argument('--training-data-dir', type=str, default=env.training_data_dir)
parser.add_argument('--batch-size', type=int, default=env.batch_size)
parser.add_argument('--model-dir', type=str, default=env.model_dir)
```

We can use [SageMaker Python SDK](https://github.com/aws/sagemaker-python-sdk) to submit ```my_training_script.py``` to
trainining using the ```TensorFlow``` estimator:

```python
# TODO (mvsusp) - allow sagemaker-python-sdk estimator to work in the scenario below
from sagemaker.tensorflow import TensorFlow

estimator = TensorFlow('train.py', role='SageMakerRole', 
                       train_instance_type='ml.p2.xlarge', train_instance_count = 1)
estimator.fit({'training': 's3://my/bucket/with/the/training/data'})
```

### Creating a new ML framework container compatible with SageMaker Training
The core training structure of SageMaker Containers is ```Training```. When training starts, e.g. 
`training.run()`, the `Training` instance executes the following steps:

1. install any required python dependencies
2. start customs metrics that will be reported to CloudWatch Logs
3. load SageMaker Training environment information, including default configuration files, directories, 
    environment variables, and hyperparameters
4. download the user script (or Python package) containing the functions required by the framework
5. start the training process
6. report success/failure
  
Let's suppose that we want to create a Keras Container that trains any Keras model provided by the **user script**.

Let's suppose that to be able to use our Containers, users need to submit a Python script containing a function with 
the following signature:  
```python
def train(training_dir, hyperparameters):
    """Return a Keras model that will be trained by our SageMaker Keras container.
    
    Args:
        - training_dir (string): the path to the directory containing the training 
            data provided by the user.
        - hyperparameters (dict[string, object]): a map containing the hyperparameters 
            provided by the user.
         
    Returns:
        - (Keras.Model): the model that will be saved in the end of the training.
    """
```

#### Implementing our Keras Framework Container
```Training``` will install the user provided script as a module, and prepare the training environment for us. All we 
have to do is implement a function, ```training_process_fn```, which implements the training process of our framework.

```training_process_fn``` takes as parameters **user module** and the **training environment**, and is responsible to 
save the trained model in end of training:
```python
import os

from sagemaker_containers import Training

def keras_framework_training_process(user_module, training_environment):
    """This function implements our Keras Framework.
    
    Args:
        - user_module (module): the python script provided by the user loaded as a Python ```module```
        - training_environment (sagemaker_containers.TrainingEnvironment): the container's 
            training environment
    """
    # retrieves the training dir from training channels
    training_dir = training_environment.channel_input_dirs['training']
    
    # retrieves the hyperparameters submitted by the user
    hps = training_environment.hyperparameters
    
    # invokes user's ```train``` function
    model = user_module.train(training_dir, hps)
    
    # saves the model in the end of training
    model.save(os.path.join(training_environment.model_dir, 'saved_model.h5'))

# Creates the training from the training process function and start training.
Training.from_train_fn(train_fn=keras_framework_training_process).run()
```

#TODO (mvsusp) - complete the example after implementing support for Docker containers.

#### An example of user provided script for training:
__train.py__
```python
import os

import keras
import numpy as np

def train(training_dir, hyperparameters):
    data = np.load(os.path.join(training_dir, hyperparameters['training_data_file']))
    x_train, y_train = data['features'], keras.utils.to_categorical(data['labels'])

    model = keras.models.Sequential()
    model.add(keras.layers.Dense(64, activation='relu', input_dim=20))
    model.add(keras.layers.Dropout(0.5))

    sgd = keras.optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
    model.compile(loss='categorical_crossentropy', optimizer=sgd, metrics=['accuracy'])

    model.fit(x_train, y_train, epochs=20, batch_size=128)
    return model
```
#### Submitting a Python script for training using our Keras container
We can use [SageMaker Python SDK](https://github.com/aws/sagemaker-python-sdk) to submit a scripts to our 
containers:

```python
# TODO (mvsusp) - allow sagemaker-python-sdk estimator to work in the scenario below
from sagemaker.estimator import Estimator

estimator = Estimator('train.py', role='SageMakerRole', 
                      train_instance_type='ml.p2.xlarge', train_instance_count = 1)
estimator.fit({'training': 's3://my/bucket/with/the/training/data'})
```

You can look at more information on how to use ```SageMaker Python SDK``` here: https://github.com/aws/sagemaker-python-sdk#byo-docker-containers-with-sagemaker-estimators.

### CODE REVIEW

**Comment by jesterhazy** [comment]
- framework code expects user code to return a model, but example user code doesn't return. actually user code isn't a function at all. 
- mxnet containers use return value from train to decide whether to call save function, and that function expects the value to be a Module/gluon network
- save_model treats model as a dict... but this would usually be a model/network of some kind, at least for mxnet users
- why do we need to set 'saved' key on model at all?

**Comment by mvsusp** [comment]
About @jesterhazy considerations:

1. user code actually returns a model. The user code example mentioned is:
```python 
def train(chanel_dirs, hps): 
    return {'trained': True, 'saved': False}
```
2. you are correct that ```train``` should return model. That is not going to change. I will add more documentation and design pictures to explain better
3. saving a model is framework specific. The example above is just a mock ML framework that saves a dict as a model
4. that is framework specific as well. I will add another example in the docs to clarify your doubts.

**Comment by jesterhazy** [comment]
i was looking at `training_module/train.py` under User Module. what's the relation between that and user script?

**Comment by mvsusp** [comment]
@jesterhazy  
Users can provide a ```def train(...):``` that returns a model.
In addition to that, users can provide a python package, that will be executed in the container with the follow command: ```python -m {module_name}```. 
The latter option allows users to execute any arbitrary script they want in the container.
Customers can use the library API in their own scripts (freedom to opt-in and use the methods that they want).

File: src/sagemaker_containers/training.py [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+import importlib
+import subprocess
+
+import sys
+
+import sagemaker_containers.training_environment as env
+
+
+class Training(object):
+    def __init__(self, train_fn=None, module_name=None):
```
Comment (iquintero) [inline src/sagemaker_containers/training.py]: ` train_fn` and `module_name` are mutually exclusive. (As inferred from the existance of  `from_train_fn()` and `from_module()`. This class needs to throw an exception if it is manually constructed and passed both.

File: src/sagemaker_containers/training.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+            else:
+                subprocess.Popen(['python', '-m', self.module_name])
+
+            self.environment.write_success_file()
+        except Exception as e:
+            self.environment.write_failure_file('Uncaught exception during training: %s' % e)
+            raise e
+
+    def import_user_module(self, code_dir, user_script='user_script'):
+        return importlib.import_module(user_script)
```
Comment (iquintero) [inline src/sagemaker_containers/training.py]: `code_dir` is not used, but it probably should right?

File: src/sagemaker_containers/training_environment.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+
+    @property
+    def input_dir(self):
+        """The base directory for training data and configuration files."""
+        return os.path.join(self.base_dir, 'input')
+
+    @property
+    def input_config_dir(self):
+        """The directory where standard SageMaker configuration files are located."""
+        return os.path.join(self.base_dir, 'input/config')
```
Comment (iquintero) [inline src/sagemaker_containers/training_environment.py]: I would rewrite this as

```python
os.path.join(self.base_dir, 'input', 'config')
```

while I think os.path.join() will figure out the / in your string correctly (even on Windows), Windows only users can be confused as to why input/config is a path.

File: src/sagemaker_containers/environment.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    important system characteristics, filesystem locations, and configuration settings.
+    """
+
+    def __init__(self):
+        self.base_dir = base_dir
+        sys.path.insert(0, self.code_dir)
+
+    @property
+    def model_dir(self):
+        return os.path.join(self.base_dir, 'model')
```
Comment (iquintero) [inline src/sagemaker_containers/environment.py]: Not a big fan of having the getter function of the property calculate its value every time. Can we move to having:

```python
def __init__(self):
    self._model_dir = os.path.join(self.base_dir, 'model')
     # ... other stuff here

@property
def model_dir(self):
    return self._model_dir
```

If someone else is strongly opposed to this Im fine discussing it though

File: test/functional/test_training.py [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+@pytest.fixture(name='opt_ml')
+def create_opt_ml(tmpdir):
+    environment.base_dir = str(tmpdir)
+    tmpdir.mkdir('output')
+    tmpdir.mkdir('code')
+    return tmpdir
+
+
+def test_training_from_train_fn(opt_ml, model_file):
```
Comment (iquintero) [inline test/functional/test_training.py]: Add a test to ensure that your class handles both a train_fn and a train_module being passed at the same time.

File: src/sagemaker_containers/training.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    def start(self):
+
+        try:
+            if self.train_fn:
+
+                user_module = self.import_user_module(self.environment.code_dir)
+
+                self.train_fn(user_module, self.environment)
+            else:
+                subprocess.Popen(['python', '-m', self.module_name])
```
Comment (iquintero) [inline src/sagemaker_containers/training.py]: this asumes that the module is importing this as a library in order to figure out where to put stuff right? or the user just hardcoded all their paths but we don't care how they do it as long as it works right?

**Review by iquintero — COMMENTED** [review]
the overall design makes sense to me, just have some comments.

**Comment by mvsusp** [comment]
@iquintero thanks for the observations. The code here was simplified for design purposes. All your comments will be addressed in the following PRs with the real code implementation. Or would you prefer to address the changes here?

**Comment by mvsusp** [comment]
@iquintero maybe I should not even push this PR since the code is not a real implementation?

File: src/sagemaker_containers/environment.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    @property
+    def code_dir(self):
+        """The directory where user-supplied code will be staged."""
+        return os.path.join(self.base_dir, 'code')
```
Comment (laurenyu) [inline src/sagemaker_containers/environment.py]: 'model' and 'code' can be constants with 'opt/ml'

File: src/sagemaker_containers/training.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+
+            self.environment.write_success_file()
+        except Exception as e:
+            self.environment.write_failure_file('Uncaught exception during training: %s' % e)
+            raise e
+
+    def import_user_module(self, code_dir, user_script='user_script'):
```
Comment (laurenyu) [inline src/sagemaker_containers/training.py]: should we be giving a default value to `user_script`? It seems a little strange to me to presume that, especially if `code_dir` is required. I'd be more inclined to follow the strategy in PythonSDK, where `entry_point` is required and `source_dir` defaults to `None`

File: src/sagemaker_containers/training_environment.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+        pass
+
+    @property
+    def channels(self):
+        """The dict of training input data channel name to directory with the input files for that channel."""
+        pass
+
+    @property
+    def channel_dirs(self):
+        """"""
```
Comment (laurenyu) [inline src/sagemaker_containers/training_environment.py]: Why bother including the docstring at this point? 😂

**Review by laurenyu — COMMENTED** [review]
are you planning a future PR that fleshes out the documentation in the README? (I'm guessing the current proposed README is just for showing what you're thinking)

also, my usual comments:
* capitalize "Python"
* import statements - I usually just ctrl/cmd-F on the [Python Standard Library list](https://docs.python.org/3/library/index.html) to check what goes in the first group.

File: src/sagemaker_containers/training.py (hunk 8 above)
Comment (mvsusp) [reply src/sagemaker_containers/training.py]: Yeah, I agree that we should not. Just did that to reduce the prototype amount of lines and focus in the overall design.

**Comment by mvsusp** [comment]
@laurenyu I agree with all your points here. I'm planning to not push this PR and just use it to gather design feedback.

File: src/sagemaker_containers/environment.py [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+
+import multiprocessing
+import os
+
+import sys
+
+base_dir = 'opt/ml'
```
Comment (owen-t) [inline src/sagemaker_containers/environment.py]: Should this be /opt/ml ?

File: src/sagemaker_containers/environment.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+#
+# or in the "license" file accompanying this file. This file is
+# distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+from __future__ import absolute_import
+
+import multiprocessing
+import os
+
```
Comment (owen-t) [inline src/sagemaker_containers/environment.py]: whitespace?

File: src/sagemaker_containers/environment.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+    @property
+    def available_cpus(self):
+        """The number of cpus available in the current container."""
+        return multiprocessing.cpu_count()
+
+    @property
+    def available_gpus(self):
+        """The number of gpus available in the current container."""
+        pass
```
Comment (owen-t) [inline src/sagemaker_containers/environment.py]: Why not provide a default here?

Try to invoke nvidia-smi (per Nacho's comment, probably not in the getter) and return the result or 0.

File: src/sagemaker_containers/training.py (hunk 1 above)
Comment (owen-t) [inline src/sagemaker_containers/training.py]: API docs! Particularly so for this class as it's a little esoteric.

File: src/sagemaker_containers/training.py [hunk 13]
```diff
@@ -0,0 +1,54 @@
...
+import subprocess
+
+import sys
+
+import sagemaker_containers.training_environment as env
+
+
+class Training(object):
+    def __init__(self, train_fn=None, module_name=None):
+        self.environment = env.TrainingEnvironment()
```
Comment (owen-t) [inline src/sagemaker_containers/training.py]: Document the constraints on these arguments

File: src/sagemaker_containers/training.py (hunk 6 above)
Comment (owen-t) [inline src/sagemaker_containers/training.py]: Popen returns immediately and successfully if the process can be created. Should this be a check_call?

File: src/sagemaker_containers/training.py [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+    @classmethod
+    def from_module(cls, module_name):
+        return cls(module_name=module_name)
+
+    def start(self):
+
+        try:
+            if self.train_fn:
+
+                user_module = self.import_user_module(self.environment.code_dir)
```
Comment (owen-t) [inline src/sagemaker_containers/training.py]: Should we run the train_fn in a subprocess as well?

File: src/sagemaker_containers/training_environment.py [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+        """The directory where standard SageMaker configuration files are located."""
+        return os.path.join(self.base_dir, 'input/config')
+
+    @property
+    def output_dir(self):
+        """The directory where training success/failure indications will be written."""
+        return os.path.join(self.base_dir, 'output')
+
+    @property
+    def resource_config(self):
```
Comment (owen-t) [inline src/sagemaker_containers/training_environment.py]: Can't we provide defaults for many of these?

File: src/sagemaker_containers/training_environment.py [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+    def channels(self):
+        """The dict of training input data channel name to directory with the input files for that channel."""
+        pass
+
+    @property
+    def channel_dirs(self):
+        """"""
+        pass
+
+    def load_training_parameters(self, fn):
```
Comment (owen-t) [inline src/sagemaker_containers/training_environment.py]: Provide doc strings for these undocumented functions

**Review by owen-t — COMMENTED** [review]
Looks good - some of the implementations are missing though - let's get them filled out!

Great stuff!

File: src/sagemaker_containers/training.py (hunk 6 above)
Comment (mvsusp) [reply src/sagemaker_containers/training.py]: When the user provides a module, we will validate it before starting the training job.
The module will be saved to the code_dir. When a new Environment instance is created, it adds the code dir to PYTHONPATH.

**Comment by mvsusp** [comment]
@jesterhazy I update the documentation and examples. Please let me know if you still have any
doubts/observations.

**Comment by mvsusp** [comment]
@laurenyu I've found this flake8 extension that inspects the import order for us, so I stop messing up the import order: https://github.com/mvsusp/sagemaker-container-support/pull/3/files#diff-b91f3d5bd63fcd17221b267e851608e8R52


🎉 🍷

File: src/sagemaker_containers/environment.py (hunk 11 above)
Comment (mvsusp) [reply src/sagemaker_containers/environment.py]: added flake8 so we won't have these error again.

File: src/sagemaker_containers/environment.py (hunk 10 above)
Comment (mvsusp) [reply src/sagemaker_containers/environment.py]: fixed.

File: src/sagemaker_containers/environment.py (hunk 4 above)
Comment (mvsusp) [reply src/sagemaker_containers/environment.py]: moved all logic to the constructor.

File: src/sagemaker_containers/environment.py (hunk 7 above)
Comment (mvsusp) [reply src/sagemaker_containers/environment.py]: created constants with just the base_names

File: src/sagemaker_containers/environment.py (hunk 12 above)
Comment (mvsusp) [reply src/sagemaker_containers/environment.py]: done

File: src/sagemaker_containers/training.py (hunk 1 above)
Comment (mvsusp) [reply src/sagemaker_containers/training.py]: @iquintero, I refactored the class to cover that scenario: https://github.com/mvsusp/sagemaker-container-support/pull/3/files#diff-35452821e8fa0a4ef79ea98492563a4aR35

Please, let me know your thoughts.

Comment (mvsusp) [reply src/sagemaker_containers/training.py]: @owen-t I added docstrings to the methods.

File: src/sagemaker_containers/training.py (hunk 13 above)
Comment (mvsusp) [reply src/sagemaker_containers/training.py]: done

File: src/sagemaker_containers/training.py (hunk 6 above)
Comment (mvsusp) [reply src/sagemaker_containers/training.py]: That's correct @owen-t. Using check call. This functional tests shows the functionality:
https://github.com/mvsusp/sagemaker-container-support/pull/3/files#diff-82c8c20785ab069f8c8d755dc15cb636R49

File: src/sagemaker_containers/training.py (hunk 2 above)
Comment (mvsusp) [reply src/sagemaker_containers/training.py]: refactored out this function: https://github.com/mvsusp/sagemaker-container-support/pull/3/files#diff-35452821e8fa0a4ef79ea98492563a4aR81

Let me know your thoughts.

File: src/sagemaker_containers/training_environment.py (hunk 3 above)
Comment (mvsusp) [reply src/sagemaker_containers/training_environment.py]: done

File: src/sagemaker_containers/training_environment.py (hunk 16 above)
Comment (mvsusp) [reply src/sagemaker_containers/training_environment.py]: done

File: test/functional/test_training.py (hunk 5 above)
Comment (mvsusp) [reply test/functional/test_training.py]: refactored the classes so we do not have the edge case anymore.

**Comment by mvsusp** [comment]
@jesterhazy @owen-t @iquintero @laurenyu I updated the readme and added tox to this PR.
The next steps will be:

- proposal design for training 
- setting up team city
- additional missing features

File: src/sagemaker_containers/environment.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+BASE_PATH = '/opt/ml'
+
+logger = logging.getLogger(__name__)
+
+_CODE_BASE_NAME = 'code'
+_MODEL_BASE_NAME = 'model'
+
+
+# TODO (mvsusp) -  create a shortcut decorator to allow get/set outside the class, e.g. env.model_dir = 'x'
+base_dir = BASE_PATH
```
Comment (jesterhazy) [inline src/sagemaker_containers/environment.py:31]: why are we exposing as model var at all? it's already a class attribute.

File: src/sagemaker_containers/training.py [hunk 18]
```diff
@@ -0,0 +1,138 @@
...
+        Args:
+            training_process_fn: a training process function that will be executed by the `Training`. Signature:
+
+                * Args:
+                    * `training_environment`(TrainingEnvironment): an instance of the `TrainingEnvironment`.
+        """
+        self.environment = env.TrainingEnvironment()
+        self.training_process_fn = training_process_fn
+
+    # TODO(mvsusp) - add update_training_environment_fn to allow frameworks to update the training env.
```
Comment (jesterhazy) [inline src/sagemaker_containers/training.py:57]: what does update mean? user can just update/attributes to the instance without adding a special function (b/c python)

File: test/functional/test_training.py [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+
+def write_user_script(tmpdir, code):
+    user_script_file = tmpdir.join('code', 'user_script.py')
+
+    user_script_file.write(code)
+    return str(user_script_file)
+
+
+def write_user_module(tmpdir, code):
+    customer_module = tmpdir.mkdir('code', 'customer_module')
```
Comment (jesterhazy) [inline test/functional/test_training.py:95]: customer_module*_path* etc.

**Review by jesterhazy — COMMENTED** [review]
looks good. added some nits. i think the readme docs will need another full pass once the design is settled.

File: src/sagemaker_containers/environment.py (hunk 17 above)
Comment (mvsusp) [reply src/sagemaker_containers/environment.py:31]: It is temporary, just to reduce the size of the PR. But I added a TODO in the top to fix it.

File: src/sagemaker_containers/training.py (hunk 18 above)
Comment (mvsusp) [reply src/sagemaker_containers/training.py:57]: for the following scenario:

```python
    
def framework_train_fn(user_module, training_env):
    model = user_module.train(training_env.channel_input_dirs, training_env.hyperparameters)
    save_model(model, model_file)

def update_training_environment_fn(training_env):
    training_env.world_size = len(training_env.hosts * training_env.num_gpus)
    training_env.rank = 2
    return training_env


Training.from_train_fn(train_fn=framework_train_fn,  
                                      environment_fn=update_training_environment_fn)
```

Put it is outside the scope this PR. We can discuss implementation details later.

File: README.md [hunk 20]
```diff
@@ -1,7 +1,188 @@
...
+
+To be able to train the script above in SageMaker, you need to provide the parameters ```--training-data-dir```,
+```--batch-size```, ```--model-dir```. ```training_environment``` makes this process very easy:
+
+```python
+import sagemaker_containers.training_environment as env
+...
+
+# TODO (mvsusp) - create env.training_data_dir and env.evaluation_data_dir helpers
+parser.add_argument('--training-data-dir', type=str, default=env.training_data_dir)
```
Comment (winstonaws) [inline README.md:67]: Does this actually work with the code as is right now?

the training_environment module doesn't have training_data_dir, batch_size, etc. fields. You have to create an instance of TrainingEnvironment, right?

It's worth thinking about whether there's value in needing to create an instance of the class. The functions could all just live in the module rather than the class. Is there a strong reason one way or the other currently?

File: src/sagemaker_containers/training.py [hunk 21]
```diff
@@ -0,0 +1,138 @@
...
+import os
+import shlex
+import subprocess
+
+import sagemaker_containers.training_environment as env
+
+_PYTHONPATH = 'PYTHONPATH'
+
+
+class Training(object):
```
Comment (winstonaws) [inline src/sagemaker_containers/training.py:25]: Two questions:

1). Can we design this so we don't lock users into this "framework"? I want to avoid the same problem we had with our original TensorFlow framework design. 

What would it look like if framework authors were implementing their train entrypoint using library functions that we provide? How much complexity would remain that we would still want to abstract away, at the cost of flexibility? I'd like to see us design that "API" first, and see some example user code in that style.

We may still want to offer something like this, which enforces a specific workflow for the entirety of training, which lets us ensure that all of our frameworks behave the same way. Even in that case, I'd argue that that should be written purely by calling those library functions.

2). For the "framework", is a class like this the best implementation? Other potential alternatives:
- A function which takes in a function/module
- A class which can be subclassed

Are there reasons for going this direction over the others?
